service/history/api/getworkflowexecutionrawhistory/api.go

```diff
@@ -0,0 +1,287 @@
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/persistence/versionhistory"
+	"go.temporal.io/server/common/rpc/interceptor"
+	"go.temporal.io/server/service/history/api"
+	"go.temporal.io/server/service/history/consts"
+	"go.temporal.io/server/service/history/events"
+	"go.temporal.io/server/service/history/shard"
+)
+
+func Invoke(
+	ctx context.Context,
+	shardContext shard.Context,
+	workflowConsistencyChecker api.WorkflowConsistencyChecker,
+	eventNotifier events.Notifier,
+	request *historyservice.GetWorkflowExecutionRawHistoryRequest,
+) (_ *historyservice.GetWorkflowExecutionRawHistoryResponse, retError error) {
+	err := validateGetWorkflowExecutionRawHistoryRequest(request)
+	if err != nil {
+		return nil, err
+	}
+	ns, err := shardContext.GetNamespaceRegistry().GetNamespaceByID(namespace.ID(request.GetNamespaceId()))
+	if err != nil {
+		return nil, err
+	}
+
+	req := request.Request
+	execution := req.Execution
+	var pageToken *tokenspb.RawHistoryContinuation
+	var targetVersionHistory *historyspb.VersionHistory
+	if req.NextPageToken == nil {
+		response, err := api.GetOrPollMutableState(
+			ctx,
+			shardContext,
+			&historyservice.GetMutableStateRequest{
+				NamespaceId: ns.ID().String(),
+				Execution:   execution,
+			},
+			workflowConsistencyChecker,
+			eventNotifier,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		targetVersionHistory, err = setRequestDefaultValueAndGetTargetVersionHistory(
+			request,
+			response.GetVersionHistories(),
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		pageToken = api.GeneratePaginationToken(request, response.GetVersionHistories())
+	} else {
+		pageToken, err = api.DeserializeRawHistoryToken(req.NextPageToken)
+		if err != nil {
+			return nil, err
+		}
+		versionHistories := pageToken.GetVersionHistories()
+		if versionHistories == nil {
+			return nil, consts.ErrInvalidVersionHistories
+		}
+		targetVersionHistory, err = setRequestDefaultValueAndGetTargetVersionHistory(
+			request,
+			versionHistories,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if err := api.ValidatePaginationToken(
+		request,
+		pageToken,
+	); err != nil {
+		return nil, err
+	}
+
+	pageSize := int(req.GetMaximumPageSize())
+	shardID := common.WorkflowIDToHistoryShard(
+		ns.ID().String(),
+		execution.GetWorkflowId(),
+		shardContext.GetConfig().NumberOfShards,
+	)
+	rawHistoryResponse, err := shardContext.GetExecutionManager().ReadRawHistoryBranch(ctx, &persistence.ReadHistoryBranchRequest{
+		BranchToken: targetVersionHistory.GetBranchToken(),
+		// GetWorkflowExecutionRawHistory is inclusive/inclusive.
+		// ReadRawHistoryBranch is inclusive/exclusive.
+		MinEventID:    pageToken.GetStartEventId(),
+		MaxEventID:    pageToken.GetEndEventId() + 1,
+		PageSize:      pageSize,
+		NextPageToken: pageToken.PersistenceToken,
+		ShardID:       shardID,
+	})
+	if err != nil {
+		if _, isNotFound := err.(*serviceerror.NotFound); isNotFound {
```

Comment:
I remember there is a service error.isNoyFoundError()

Reply:
Sure. Will use that instead.